drivers/sdhc/sdhc_esp32.c

```diff
@@ -0,0 +1,1552 @@
+
+	/* Post init settings */
+
+	for (int i = 0; i < cfg->slots; i++) {
+		ret = sdmmc_host_set_card_clk(sdio_hw, cfg->slots_cfg[i].num,
+					      data->bus_clock / 1000);
+
+		if (ret != 0) {
+			LOG_ERR("Error configuring card clock on slot %d", cfg->slots_cfg[i].num);
+			return err_esp2zep(ret);
+		}
+
+		ret = sdmmc_host_set_bus_width(sdio_hw, cfg->slots_cfg[i].num,
+					       SDMMC_SLOT_WIDTH_DEFAULT);
+
+		if (ret != 0) {
+			LOG_ERR("Error configuring bus width on slot %d", cfg->slots_cfg[i].num);
+			return err_esp2zep(ret);
+		}
+
+		/* Data init */
+		data->slots_data[i].bus_width = SDMMC_SLOT_WIDTH_DEFAULT;
+		data->slots_data[i].power_mode = SDHC_POWER_ON;
+		data->slots_data[i].timing = SDHC_TIMING_LEGACY;
+	}
+
+	return 0;
+}
+
+static const struct sdhc_driver_api sdhc_api = {.reset = sdhc_esp32_reset,
+						.request = sdhc_esp32_request,
+						.set_io = sdhc_esp32_set_io,
+						.get_card_present = sdhc_esp32_get_card_present,
+						.card_busy = sdhc_esp32_card_busy,
+						.get_host_props = sdhc_esp32_get_host_props};
+
+#define SLOT_CONFIG(id)                                                                    \
+	{                                                                                      \
+		.num = DT_PROP(id, reg),                                                           \
+		.bus_width_cfg = DT_PROP(id, bus_width),                                           \
+		.pcfg = PINCTRL_DT_DEV_CONFIG_GET(id),                                             \
+		.pwr_gpio = GPIO_DT_SPEC_GET_OR(id, pwr_gpios, {0}),                               \
+		.clk_pin = DT_PROP_OR(id, clk_pin, GPIO_NUM_NC),                                   \
+		.cmd_pin = DT_PROP_OR(id, cmd_pin, GPIO_NUM_NC),                                   \
+		.d0_pin = DT_PROP_OR(id, d0_pin, GPIO_NUM_NC),                                     \
+		.d1_pin = DT_PROP_OR(id, d1_pin, GPIO_NUM_NC),                                     \
+		.d2_pin = DT_PROP_OR(id, d2_pin, GPIO_NUM_NC),                                     \
+		.d3_pin = DT_PROP_OR(id, d3_pin, GPIO_NUM_NC),                                     \
+		.power_delay_ms = DT_PROP_OR(id, power_delay_ms, 0),                               \
+	},
+
+#define SDHC_ESP32_INIT(n)                                                                     \
+                                                                                               \
+	K_MSGQ_DEFINE(sdhc##n##_queue, sizeof(struct sdmmc_event), SDMMC_EVENT_QUEUE_LENGTH, 1);   \
+                                                                                               \
+	DT_FOREACH_CHILD_STATUS_OKAY_SEP(DT_INST_CHILD(n, slots), PINCTRL_DT_DEFINE, ";");         \
+                                                                                               \
+	static const struct sdhc_esp32_slot_config sdhc_esp32_##n##_slot_config[] = {              \
+		DT_FOREACH_CHILD_STATUS_OKAY(DT_INST_CHILD(n, slots), SLOT_CONFIG)};               \
```

Comment:
We do not want to have slot configurations defined by iterating over each child node. Rather, we want to have each *slot* be iterated over by the `DT_INST_FOREACH_STATUS_OKAY` macro, and then have properties that are shared by each child node be read using the `DT_PARENT` macro

Reply:
Thank you. Is the current implementation closer to what you describe?

Reply:
Yes it is- thank you very much for reworking this, it looks much better :)